File: node/src/crossChain.ts
// ──────────────────────────────────────────────────────────────────────────
// Cross-chain orchestration via LiFi.
//
// We do NOT depend on @lifi/sdk. We call their public REST API (li.quest/v1)
// directly so the SDK footprint stays tiny and the flow is easy to mock in
// tests. The same approach used by their own quote endpoint examples.
//
// Architectural note — `Obsidian/21 - Unified Agent Economy.md` non-goal:
//   "We do not move funds ourselves. Cross-chain settlement is delegated to
//    established bridges (LiFi, Jupiter, Wormhole) — we orchestrate the
//    sequence and the agent signs every step."
//
// This file is the orchestration layer for EVM↔EVM. Solana↔EVM will live
// in `crossChainSolana.ts` (Phase 1.5b: Wormhole/deBridge + Jupiter swap).
// ──────────────────────────────────────────────────────────────────────────
import type { PublicClient, WalletClient } from "viem";
import { AiFinPayError } from "./errors.js";

const LIFI_API = "https://li.quest/v1";

// ── Supported EVM chains for cross-chain settlement ──────────────────────
// EVM chain IDs (canonical, used by both viem and LiFi).
export const EVM_CHAINS = {
  ethereum:  1,
  polygon:   137,
  bsc:       56,
  arbitrum:  42161,
  optimism:  10,
  base:      8453,
} as const;

export type EvmChainName = keyof typeof EVM_CHAINS;

// USDC token addresses per chain. Native (Circle CCTP) variant where it
// exists; bridged USDC.e listed in `USDC_BRIDGED` for legacy compatibility.
export const USDC_NATIVE: Record<EvmChainName, `0x${string}`> = {
  ethereum: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
  polygon:  "0x3c499c542cEF5E3811e1192ce70d8cC03d5c3359",
  bsc:      "0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d", // Binance-Peg BSC-USD; closest analogue
  arbitrum: "0xaf88d065e77c8cC2239327C5EDb3A432268e5831",
  optimism: "0x0b2C639c533813f4Aa9D7837CAf62653d097Ff85",
  base:     "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
};

export const USDC_BRIDGED: Partial<Record<EvmChainName, `0x${string}`>> = {
  polygon:  "0x2791Bca1f2de4661ED88A30C99A7a9449Aa84174", // USDC.e PoS bridged (deprecated, kept for legacy merchants)
  arbitrum: "0xFF970A61A04b1cA14834A43f5dE4533eBDDB5CC8", // USDC.e (deprecated)
};

// ── Quote types ──────────────────────────────────────────────────────────

export interface BridgeQuoteOptions {
  fromChain:    EvmChainName;
  toChain:      EvmChainName;
  fromToken:    `0x${string}`;         // ERC-20 address on source chain
  toToken:      `0x${string}`;         // ERC-20 address on dest chain
  fromAmount:   string;                 // base units as string (USDC is 6 decimals)
  fromAddress:  `0x${string}`;          // agent's EVM address
  toAddress?:   `0x${string}`;          // defaults to fromAddress
  slippage?:    number;                 // 0.005 = 0.5%
  // LiFi accepts `integrator` for analytics + revenue share with project
  // partners. We pass "aifinpay" so all our agent-driven volume is tagged.
  integrator?:  string;
}

// Subset of LiFi's quote response shape. Their full schema is large; we
// only type-pin the fields we read. Unknown fields stay accessible via
// `raw_quote.<anything>` for partners who want the full route detail.
export interface BridgeQuote {
  // Human-readable summary
  from: { chain: EvmChainName; token: `0x${string}`; amount: string };
  to:   { chain: EvmChainName; token: `0x${string}`; amount: string; amount_min: string };
  fees: { bridge_usd: number; gas_usd: number; total_usd: number };
  eta_seconds:  number;
  bridge_tool:  string; // "stargate", "across", "circle-cctp", etc.
  // Raw LiFi response. Pass straight to `bridgeExecute()`.
  raw_quote:    unknown;
}

interface LifiQuoteResponse {
  estimate: {
    fromAmount:    string;
    toAmount:      string;
    toAmountMin:   string;
    feeCosts?:     Array<{ amountUSD?: string; name?: string }>;
    gasCosts?:     Array<{ amountUSD?: string }>;
    executionDuration?: number;
  };
  transactionRequest: {
    to:        `0x${string}`;
    data:      `0x${string}`;
    value:     string;            // hex (e.g. "0x0")
    gasLimit:  string;            // hex
    chainId:   number;
  };
  tool:        string;
  toolDetails?: { name?: string };
  action: {
    fromChainId: number;
    toChainId:   number;
    fromToken:   { address: `0x${string}` };
    toToken:     { address: `0x${string}` };
  };
}

// ── Quote ────────────────────────────────────────────────────────────────

export async function bridgeQuote(opts: BridgeQuoteOptions): Promise<BridgeQuote> {
  const fromChainId = EVM_CHAINS[opts.fromChain];
  const toChainId   = EVM_CHAINS[opts.toChain];

  const url = new URL(`${LIFI_API}/quote`);
  url.searchParams.set("fromChain",   String(fromChainId));
  url.searchParams.set("toChain",     String(toChainId));
  url.searchParams.set("fromToken",   opts.fromToken);
  url.searchParams.set("toToken",     opts.toToken);
  url.searchParams.set("fromAmount",  opts.fromAmount);
  url.searchParams.set("fromAddress", opts.fromAddress);
  if (opts.toAddress) url.searchParams.set("toAddress", opts.toAddress);
  if (opts.slippage !== undefined) url.searchParams.set("slippage", String(opts.slippage));
  url.searchParams.set("integrator", opts.integrator ?? "aifinpay");

  const r = await fetch(url.toString());
  if (!r.ok) {
    const detail = await r.text().catch(() => "<unreadable>");
    throw new AiFinPayError(
      `bridgeQuote: LiFi /quote returned ${r.status} for ${opts.fromChain}→${opts.toChain}: ${detail.slice(0, 300)}`,
    );
  }
  const j = (await r.json()) as LifiQuoteResponse;

  const bridgeUsd = (j.estimate.feeCosts ?? []).reduce((s, f) => s + Number(f.amountUSD ?? 0), 0);
  const gasUsd    = (j.estimate.gasCosts ?? []).reduce((s, g) => s + Number(g.amountUSD ?? 0), 0);

  return {
    from: { chain: opts.fromChain, token: opts.fromToken, amount: j.estimate.fromAmount },
    to:   {
      chain:      opts.toChain,
      token:      opts.toToken,
      amount:     j.estimate.toAmount,
      amount_min: j.estimate.toAmountMin,
    },
    fees: { bridge_usd: bridgeUsd, gas_usd: gasUsd, total_usd: bridgeUsd + gasUsd },
    eta_seconds: j.estimate.executionDuration ?? 0,
    bridge_tool: j.toolDetails?.name ?? j.tool,
    raw_quote:   j,
  };
}

// ── Execute ──────────────────────────────────────────────────────────────

export interface BridgeReceipt {
  source_tx:     `0x${string}`;
  source_chain:  EvmChainName;
  dest_chain:    EvmChainName;
  bridge_tool:   string;
  status:        "submitted" | "pending" | "done" | "failed";
  // dest_tx is populated only after polling status until completion.
  // Call `bridgeWaitForArrival()` separately if you need it inline.
  dest_tx?:      string;
}

/**
 * Submit the source-chain transaction returned by `bridgeQuote()`.
 * Caller supplies a viem walletClient connected to the SOURCE chain (the
 * agent's wallet on `opts.fromChain`). We don't switch chains for the
 * caller — that's an explicit choice to keep this primitive small.
 *
 * The agent signs and submits. AiFinPay never touches the funds.
 */
export async function bridgeExecute(
  quote:          BridgeQuote,
  walletClient:   WalletClient,
  publicClient:   PublicClient,
): Promise<BridgeReceipt> {
  const raw = quote.raw_quote as LifiQuoteResponse;
  const tx  = raw.transactionRequest;
  if (!tx) {
    throw new AiFinPayError(
      `bridgeExecute: quote has no transactionRequest — likely returned by /quote/toAmount which we don't use`,
    );
  }

  // Sanity-check that the wallet client is on the source chain.
  const chainId = await walletClient.getChainId();
  if (chainId !== tx.chainId) {
    throw new AiFinPayError(
      `bridgeExecute: walletClient is on chain ${chainId}, quote requires ${tx.chainId} (${quote.from.chain})`,
    );
  }

  const account = walletClient.account;
  if (!account) {
    throw new AiFinPayError(`bridgeExecute: walletClient has no account — pass account: at creation`);
  }

  const hash = await walletClient.sendTransaction({
    account,
    to:       tx.to,
    data:     tx.data,
    value:    BigInt(tx.value || "0x0"),
    gas:      tx.gasLimit ? BigInt(tx.gasLimit) : undefined,
    chain:    null,
  });

  // Wait for source-chain inclusion only. Dest-chain arrival is async —
  // caller polls via `bridgeWaitForArrival(hash)` if they care.
  const receipt = await publicClient.waitForTransactionReceipt({ hash });
  if (receipt.status !== "success") {
    return {
      source_tx:    hash,
      source_chain: quote.from.chain,
      dest_chain:   quote.to.chain,
      bridge_tool:  quote.bridge_tool,
      status:       "failed",
    };
  }

  return {
    source_tx:    hash,
    source_chain: quote.from.chain,
    dest_chain:   quote.to.chain,
    bridge_tool:  quote.bridge_tool,
    status:       "submitted",
  };
}

// ── Status polling ───────────────────────────────────────────────────────

/**
 * Poll LiFi's /v1/status for cross-chain arrival. Stargate/Across typically
 * finalise in 30s-3min; Circle CCTP can take 15-25min on Polygon side.
 *
 * Returns once status is "DONE" or "FAILED", or throws on timeout.
 */
export async function bridgeWaitForArrival(
  sourceTxHash:  `0x${string}`,
  opts: {
    pollIntervalMs?: number;
    timeoutMs?:      number;
  } = {},
): Promise<{ status: "done" | "failed"; dest_tx?: string; raw: unknown }> {
  const pollMs = opts.pollIntervalMs ?? 5000;
  const timeoutMs = opts.timeoutMs   ?? 30 * 60 * 1000; // 30 min ceiling
  const deadline = Date.now() + timeoutMs;

  while (Date.now() < deadline) {
    const url = `${LIFI_API}/status?txHash=${sourceTxHash}`;
    const r = await fetch(url);
    if (r.ok) {
      const j = (await r.json()) as {
        status?:    string;
        receiving?: { txHash?: string };
      };
      if (j.status === "DONE") {
        return { status: "done", dest_tx: j.receiving?.txHash, raw: j };
      }
      if (j.status === "FAILED" || j.status === "INVALID") {
        return { status: "failed", raw: j };
      }
    }
    await new Promise((resolve) => setTimeout(resolve, pollMs));
  }
  throw new AiFinPayError(
    `bridgeWaitForArrival: timeout after ${timeoutMs}ms — source tx ${sourceTxHash} did not finalise on dest`,
  );
}

File: python/aifinpay/cross_chain.py
"""
Cross-chain orchestration via LiFi (Phase 1.5a — EVM↔EVM).

Mirrors `aifinpay-sdk/node/src/crossChain.ts`. We do NOT depend on a
LiFi-specific SDK — we call the public REST API at https://li.quest/v1
directly with `requests`, so the SDK footprint stays tiny and the flow
is easy to mock in tests.

Architectural note — Obsidian/21 - Unified Agent Economy.md non-goal:
    "We do not move funds ourselves. Cross-chain settlement is delegated
     to established bridges (LiFi, Jupiter, Wormhole) — we orchestrate
     the sequence and the agent signs every step."

This file is the orchestration layer for EVM↔EVM. Solana↔EVM will live
in `cross_chain_solana.py` (Phase 1.5b: Wormhole/deBridge + Jupiter swap).
"""
from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Any, Dict, Optional

import requests

from .errors import AiFinPayError

LIFI_API = "https://li.quest/v1"

# ── Supported EVM chains for cross-chain settlement ──────────────────────
# EVM chain IDs (canonical, used by both web3.py and LiFi). Ported verbatim
# from the Node SDK.
EVM_CHAINS: Dict[str, int] = {
    "ethereum": 1,
    "polygon":  137,
    "bsc":      56,
    "arbitrum": 42161,
    "optimism": 10,
    "base":     8453,
}

# USDC token addresses per chain. Native (Circle CCTP) variant where it
# exists; bridged USDC.e listed in `USDC_BRIDGED` for legacy compatibility.
USDC_NATIVE: Dict[str, str] = {
    "ethereum": "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
    "polygon":  "0x3c499c542cEF5E3811e1192ce70d8cC03d5c3359",
    "bsc":      "0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d",
    "arbitrum": "0xaf88d065e77c8cC2239327C5EDb3A432268e5831",
    "optimism": "0x0b2C639c533813f4Aa9D7837CAf62653d097Ff85",
    "base":     "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
}

USDC_BRIDGED: Dict[str, str] = {
    "polygon":  "0x2791Bca1f2de4661ED88A30C99A7a9449Aa84174",
    "arbitrum": "0xFF970A61A04b1cA14834A43f5dE4533eBDDB5CC8",
}


# ── Dataclasses ─────────────────────────────────────────────────────────


@dataclass
class BridgeQuoteFrom:
    chain:  str
    token:  str
    amount: str


@dataclass
class BridgeQuoteTo:
    chain:      str
    token:      str
    amount:     str
    amount_min: str


@dataclass
class BridgeQuoteFees:
    bridge_usd: float
    gas_usd:    float
    total_usd:  float


@dataclass
class BridgeQuote:
    """Subset of LiFi's quote response; full payload in `raw_quote`."""
    from_:       BridgeQuoteFrom
    to:          BridgeQuoteTo
    fees:        BridgeQuoteFees
    eta_seconds: int
    bridge_tool: str
    raw_quote:   Dict[str, Any] = field(default_factory=dict)


@dataclass
class BridgeReceipt:
    source_tx:    str
    source_chain: str
    dest_chain:   str
    bridge_tool:  str
    status:       str   # "submitted" | "pending" | "done" | "failed"
    dest_tx:      Optional[str] = None


# ── Quote ───────────────────────────────────────────────────────────────


def bridge_quote(
    *,
    from_chain:   str,
    to_chain:     str,
    from_token:   str,
    to_token:     str,
    from_amount:  str,
    from_address: str,
    to_address:   Optional[str] = None,
    slippage:     Optional[float] = None,
    integrator:   str = "aifinpay",
    timeout:      float = 30.0,
) -> BridgeQuote:
    """Fetch a cross-chain quote from LiFi.

    `from_amount` is in base units as a string (USDC = 6 decimals, so $1.00
    is "1000000"). Use the convenience method on `AiFinPayAgent.bridge_quote`
    to pass `amount_usdc` directly.

    `integrator` is forwarded to LiFi for analytics + revenue share with
    project partners; defaults to "aifinpay" so all our agent-driven volume
    is tagged.
    """
    if from_chain not in EVM_CHAINS:
        raise AiFinPayError(f"bridge_quote: unknown from_chain {from_chain!r}")
    if to_chain not in EVM_CHAINS:
        raise AiFinPayError(f"bridge_quote: unknown to_chain {to_chain!r}")

    params: Dict[str, Any] = {
        "fromChain":   EVM_CHAINS[from_chain],
        "toChain":     EVM_CHAINS[to_chain],
        "fromToken":   from_token,
        "toToken":     to_token,
        "fromAmount":  from_amount,
        "fromAddress": from_address,
        "integrator":  integrator,
    }
    if to_address is not None:
        params["toAddress"] = to_address
    if slippage is not None:
        params["slippage"] = slippage

    r = requests.get(f"{LIFI_API}/quote", params=params, timeout=timeout)
    if not r.ok:
        detail = (r.text or "<unreadable>")[:300]
        raise AiFinPayError(
            f"bridge_quote: LiFi /quote returned {r.status_code} for "
            f"{from_chain}->{to_chain}: {detail}"
        )
    j = r.json()

    estimate = j.get("estimate", {}) or {}
    fee_costs = estimate.get("feeCosts") or []
    gas_costs = estimate.get("gasCosts") or []
    bridge_usd = sum(float(f.get("amountUSD", 0) or 0) for f in fee_costs)
    gas_usd    = sum(float(g.get("amountUSD", 0) or 0) for g in gas_costs)

    tool_details = j.get("toolDetails") or {}
    bridge_tool = tool_details.get("name") or j.get("tool", "")

    return BridgeQuote(
        from_=BridgeQuoteFrom(
            chain=from_chain,
            token=from_token,
            amount=str(estimate.get("fromAmount", from_amount)),
        ),
        to=BridgeQuoteTo(
            chain=to_chain,
            token=to_token,
            amount=str(estimate.get("toAmount", "0")),
            amount_min=str(estimate.get("toAmountMin", "0")),
        ),
        fees=BridgeQuoteFees(
            bridge_usd=bridge_usd,
            gas_usd=gas_usd,
            total_usd=bridge_usd + gas_usd,
        ),
        eta_seconds=int(estimate.get("executionDuration") or 0),
        bridge_tool=bridge_tool,
        raw_quote=j,
    )


# ── Execute ─────────────────────────────────────────────────────────────


def bridge_execute(
    quote:           BridgeQuote,
    *,
    evm_account:     Any,          # eth_account.LocalAccount
    web3:            Any,          # web3.Web3 instance connected to source chain
) -> BridgeReceipt:
    """Submit the source-chain transaction returned by `bridge_quote()`.

    Signs and submits the LiFi-returned `transactionRequest` on the SOURCE
    chain with the agent's EVM key. Returns once source-side inclusion is
    confirmed; dest-side arrival is async — call `bridge_wait_for_arrival(
    receipt.source_tx)` if you need to block on it.

    The agent signs and submits. AiFinPay never touches the funds.

    Note: web3.py is a required dependency of this SDK (declared in
    pyproject.toml), so this method works out of the box. The caller must
    pass a Web3 instance already pointed at the source chain's RPC — we
    don't switch chains for the caller, matching the Node SDK's explicit
    choice to keep this primitive small.
    """
    raw = quote.raw_quote or {}
    tx = raw.get("transactionRequest")
    if not tx:
        raise AiFinPayError(
            "bridge_execute: quote has no transactionRequest — likely returned "
            "by /quote/toAmount which we don't use"
        )

    # Sanity-check the Web3 instance is on the source chain.
    try:
        chain_id = web3.eth.chain_id
    except Exception as e:
        raise AiFinPayError(f"bridge_execute: web3 not connected: {e}") from e
    expected_chain_id = int(tx.get("chainId", EVM_CHAINS[quote.from_.chain]))
    if chain_id != expected_chain_id:
        raise AiFinPayError(
            f"bridge_execute: web3 is on chain {chain_id}, quote requires "
            f"{expected_chain_id} ({quote.from_.chain})"
        )

    # Build the tx. LiFi returns hex strings for value / gasLimit.
    value = int(tx.get("value", "0x0"), 16) if isinstance(tx.get("value"), str) else int(tx.get("value") or 0)
    gas_limit = (
        int(tx["gasLimit"], 16)
        if isinstance(tx.get("gasLimit"), str)
        else int(tx.get("gasLimit") or 0)
    )

    nonce = web3.eth.get_transaction_count(evm_account.address)
    tx_payload: Dict[str, Any] = {
        "from":     evm_account.address,
        "to":       tx["to"],
        "data":     tx["data"],
        "value":    value,
        "nonce":    nonce,
        "chainId":  expected_chain_id,
        "gasPrice": web3.eth.gas_price,
    }
    if gas_limit:
        tx_payload["gas"] = gas_limit
    else:
        try:
            tx_payload["gas"] = int(web3.eth.estimate_gas(tx_payload) * 1.2)
        except Exception:
            tx_payload["gas"] = 500_000  # safe default for bridge txs

    signed = web3.eth.account.sign_transaction(tx_payload, evm_account.key)
    tx_hash = web3.eth.send_raw_transaction(signed.raw_transaction)
    receipt = web3.eth.wait_for_transaction_receipt(tx_hash, timeout=180)
    tx_hash_hex = tx_hash.hex() if hasattr(tx_hash, "hex") else str(tx_hash)
    if not tx_hash_hex.startswith("0x"):
        tx_hash_hex = "0x" + tx_hash_hex

    if receipt.status != 1:
        return BridgeReceipt(
            source_tx=tx_hash_hex,
            source_chain=quote.from_.chain,
            dest_chain=quote.to.chain,
            bridge_tool=quote.bridge_tool,
            status="failed",
        )
    return BridgeReceipt(
        source_tx=tx_hash_hex,
        source_chain=quote.from_.chain,
        dest_chain=quote.to.chain,
        bridge_tool=quote.bridge_tool,
        status="submitted",
    )


# ── Status polling ──────────────────────────────────────────────────────


def bridge_wait_for_arrival(
    source_tx_hash:   str,
    *,
    poll_interval_ms: int = 5000,
    timeout_ms:       int = 30 * 60 * 1000,
    request_timeout:  float = 30.0,
) -> Dict[str, Any]:
    """Poll LiFi's /v1/status for cross-chain arrival.

    Stargate/Across typically finalise in 30s-3min; Circle CCTP can take
    15-25min on Polygon side. Returns a dict with::

        {"status": "done" | "failed", "dest_tx": str | None, "raw": ...}

    Raises `AiFinPayError` on timeout.
    """
    poll_s = poll_interval_ms / 1000.0
    deadline = time.time() + (timeout_ms / 1000.0)

    while time.time() < deadline:
        try:
            r = requests.get(
                f"{LIFI_API}/status",
                params={"txHash": source_tx_hash},
                timeout=request_timeout,
            )
            if r.ok:
                j = r.json()
                status = j.get("status")
                if status == "DONE":
                    receiving = j.get("receiving") or {}
                    return {
                        "status":  "done",
                        "dest_tx": receiving.get("txHash"),
                        "raw":     j,
                    }
                if status in ("FAILED", "INVALID"):
                    return {"status": "failed", "dest_tx": None, "raw": j}
        except requests.RequestException:
            # Transient network errors are swallowed; we'll retry on the
            # next poll. A persistent outage hits the timeout below.
            pass
        time.sleep(poll_s)

    raise AiFinPayError(
        f"bridge_wait_for_arrival: timeout after {timeout_ms}ms — "
        f"source tx {source_tx_hash} did not finalise on dest"
    )
